File: kotlin/payment-service/domain/src/main/kotlin/com/example/payment/domain/model/Payment.kt
package com.example.payment.domain.model

import com.example.payment.domain.PaymentStatus
import com.example.payment.domain.exception.InvalidPaymentStateException
import java.math.BigDecimal
import java.time.Instant

class Payment private constructor(
    val id: PaymentId,
    val orderId: String,
    val amount: Money,
    status: PaymentStatus,
    val createdAt: Instant,
    updatedAt: Instant,
) {
    var status: PaymentStatus = status
        private set

    var updatedAt: Instant = updatedAt
        private set

    fun capture() {
        transitTo(PaymentStatus.CAPTURED)
    }

    fun refund() {
        transitTo(PaymentStatus.REFUNDED)
    }

    private fun transitTo(target: PaymentStatus) {
        if (!status.canTransitionTo(target)) {
            throw InvalidPaymentStateException(status, target)
        }
        status = target
        updatedAt = Instant.now()
    }

    companion object {
        fun authorize(
            paymentId: PaymentId,
            orderId: String,
            amount: BigDecimal,
        ): Payment {
            val now = Instant.now()
            return Payment(
                id = paymentId,
                orderId = orderId,
                amount = Money(amount),
                status = PaymentStatus.AUTHORIZED,
                createdAt = now,
                updatedAt = now,
            )
        }

        fun reconstitute(
            paymentId: PaymentId,
            orderId: String,
            amount: Money,
            status: PaymentStatus,
            createdAt: Instant,
            updatedAt: Instant,
        ): Payment = Payment(
            id = paymentId,
            orderId = orderId,
            amount = amount,
            status = status,
            createdAt = createdAt,
            updatedAt = updatedAt,
        )
    }
}

File: go/payment-service/internal/payment/payment.go
package payment

import (
	"fmt"
	"time"

	"github.com/google/uuid"
)

type Status string

const (
	StatusAuthorized Status = "AUTHORIZED"
	StatusCaptured   Status = "CAPTURED"
	StatusRefunded   Status = "REFUNDED"
	StatusFailed     Status = "FAILED"
)

var validTransitions = map[Status][]Status{
	StatusAuthorized: {StatusCaptured, StatusRefunded},
	StatusCaptured:   {StatusRefunded},
	StatusRefunded:   {},
	StatusFailed:     {},
}

func (s Status) CanTransitionTo(target Status) bool {
	for _, t := range validTransitions[s] {
		if t == target {
			return true
		}
	}
	return false
}

type Payment struct {
	ID        string
	OrderID   string
	Amount    int64
	Status    Status
	CreatedAt time.Time
	UpdatedAt time.Time
}

func NewPayment(orderID string, amount int64) (*Payment, error) {
	if amount < 0 {
		return nil, fmt.Errorf("amount must be >= 0, got %d", amount)
	}
	now := time.Now().UTC()
	return &Payment{
		ID:        uuid.New().String(),
		OrderID:   orderID,
		Amount:    amount,
		Status:    StatusAuthorized,
		CreatedAt: now,
		UpdatedAt: now,
	}, nil
}

func (p *Payment) Capture() error {
	if !p.Status.CanTransitionTo(StatusCaptured) {
		return &InvalidStatusError{
			Current: p.Status,
			Target:  StatusCaptured,
		}
	}
	p.Status = StatusCaptured
	p.UpdatedAt = time.Now().UTC()
	return nil
}

func (p *Payment) Refund() error {
	if !p.Status.CanTransitionTo(StatusRefunded) {
		return &InvalidStatusError{
			Current: p.Status,
			Target:  StatusRefunded,
		}
	}
	p.Status = StatusRefunded
	p.UpdatedAt = time.Now().UTC()
	return nil
}

type FailedError struct {
	OrderID string
	Reason  string
}

func (e *FailedError) Error() string {
	return fmt.Sprintf("Payment failed for order %s: %s", e.OrderID, e.Reason)
}

type InvalidStatusError struct {
	Current Status
	Target  Status
}

func (e *InvalidStatusError) Error() string {
	return fmt.Sprintf("Cannot transition from %s to %s", e.Current, e.Target)
}
